### PR TITLE
hard: add tower of hanoi

### PR DESCRIPTION
**Pull request by yogo1212**

### CODE REVIEW

File: Python/hard/hanoi/solution.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+
+from os import system
+import time
+import sys
+
+def solve(towers, a, b, c, i):
```
Comment (gandie) [inline Python/hard/hanoi/solution.py]: Please use more precise variable names instead of `a, b, c`

Comment (gandie) [reply Python/hard/hanoi/solution.py]: ... and i :rofl:

Comment (gandie) [reply Python/hard/hanoi/solution.py]: ... and add a nice docstring explaining the function

File: Python/hard/hanoi/solution.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+def display(towers, max_width = None):
+	if max_width is None:
+		# the biggest disc must be at the bottom
+		bottom_widths = [
+			tower[0] if len(tower) > 0 else 0
+			for tower in towers
+		]
+
+		max_width = max(*bottom_widths)
```
Comment (gandie) [inline Python/hard/hanoi/solution.py]: No `*` needed here, plz remove

File: Python/hard/hanoi/solution.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+	print()
+	for f in range(max_width - 1, -1, -1):
+		floors = [floor_string(towers[t], max_width, f) for t in range(3)]
```
Comment (gandie) [inline Python/hard/hanoi/solution.py]: Plz add some line breaks here for readability, like list comprehension above. In general one can write:

```
something = [
    operation(item)
    for item in items
    if some_filter(item)
]
```

File: Python/hard/hanoi/solution.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+if __name__ == '__main__':
+	towers = [[],[],[]]
+
+	discs = 5
+	if len(sys.argv) > 1:
+		discs = int(sys.argv[1])
+
+	assert discs < 20, "disc count must be below 20"
+
+	# initialize towerz
```
Comment (gandie) [inline Python/hard/hanoi/solution.py]:
```
initialize towerz, from bottom to top
```

File: Python/hard/hanoi/solution.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+	buf = "▄" * tower[pos]
+	rem = width - tower[pos]
+	if rem % 2 == 1:
+		buf = '▗' + buf[1:] + "▖"
+
+	cushion = " " * int(rem / 2)
+	return cushion + buf + cushion
+
+def display(towers, max_width = None):
```
Comment (gandie) [inline Python/hard/hanoi/solution.py:36]: Add a short docstring here explaining this function

File: Python/hard/hanoi/solution.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		return
+
+	solve(towers, a, c, b, i - 1)
+
+	towers[c].append(towers[a].pop())
+	display(towers)
+
+	solve(towers, b, a, c, i - 1)
+
+def floor_string(tower, width, pos):
```
Comment (gandie) [inline Python/hard/hanoi/solution.py]: Add a short docstring here explaining this function.

**Review by gandie — CHANGES_REQUESTED** [review]
nice one! some minor changes and additions requested :wink:

**Comment by gandie** [comment]
And please make sure indendation is consistent → 4 spaces

File: Python/hard/hanoi/solution.py (hunk 1 above)
Comment (yogo1212) [reply Python/hard/hanoi/solution.py]: docstring fine?

File: Python/hard/hanoi/solution.py (hunk 4 above)
Comment (yogo1212) [reply Python/hard/hanoi/solution.py]: like this?

**Review by gandie — APPROVED** [review]
nice one, lgtm!

**Pull request merged**